shell/src/main/java/org/apache/zeppelin/shell/TerminalInterpreter.java

```diff
@@ -0,0 +1,221 @@
+import org.apache.zeppelin.interpreter.InterpreterResult.Code;
+import org.apache.zeppelin.interpreter.InterpreterResultMessageOutput;
+import org.apache.zeppelin.interpreter.KerberosInterpreter;
+import org.apache.zeppelin.interpreter.remote.RemoteInterpreterUtils;
+import org.apache.zeppelin.interpreter.thrift.InterpreterCompletion;
+import org.apache.zeppelin.scheduler.Scheduler;
+import org.apache.zeppelin.scheduler.SchedulerFactory;
+import org.apache.zeppelin.shell.terminal.TerminalManager;
+import org.apache.zeppelin.shell.terminal.TerminalThread;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.URL;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Properties;
+
+/**
+ * Shell interpreter for Zeppelin.
+ */
+public class TerminalInterpreter extends KerberosInterpreter {
+  private static final Logger LOGGER = LoggerFactory.getLogger(TerminalInterpreter.class);
+
+  public TerminalInterpreter(Properties property) {
+    super(property);
+  }
+
+  private TerminalThread terminalThread = null;
+
+  private InterpreterContext intpContext;
+
+  private int terminalPort = 0;
+
+  // terminal web socket status
+  // ui_templates/terminal-dashboard.jinja
+  public static final String TERMINAL_SOCKET_STATUS = "TERMINAL_SOCKET_STATUS";
+  public static final String TERMINAL_SOCKET_CONNECT = "TERMINAL_SOCKET_CONNECT";
+  public static final String TERMINAL_SOCKET_CLOSE = "TERMINAL_SOCKET_CLOSE";
+  public static final String TERMINAL_SOCKET_ERROR = "TERMINAL_SOCKET_ERROR";
+
+  @Override
+  public void open() {
+    super.open();
+  }
+
+  private void setParagraphConfig() {
+    intpContext.getConfig().put("editorHide", true);
+    intpContext.getConfig().put("title", false);
+  }
+
+  @Override
+  public void close() {
+    intpContext.getAngularObjectRegistry().add(TERMINAL_SOCKET_STATUS, TERMINAL_SOCKET_CLOSE,
+        intpContext.getNoteId(), intpContext.getParagraphId());
+    if (null != terminalThread) {
+      TerminalManager.getInstance().cleanIntpContext(intpContext.getNoteId());
+      terminalThread.stopRunning();
+    }
+    super.close();
+  }
+
+  @Override
+  public BaseZeppelinContext getZeppelinContext() {
+    return null;
+  }
+
+  @Override
+  public InterpreterResult internalInterpret(String cmd, InterpreterContext context) {
+    this.intpContext = context;
+    TerminalManager.getInstance().setInterpreterContext(context);
+
+    if (null == terminalThread) {
+      try {
+        terminalPort = RemoteInterpreterUtils.findRandomAvailablePortOnAllLocalInterfaces();
+        terminalThread = new TerminalThread(terminalPort);
+        terminalThread.start();
+      } catch (IOException e) {
+        LOGGER.error(e.getMessage(), e);
+        return new InterpreterResult(Code.ERROR, e.getMessage());
+      }
+
+      for (int i = 0; i < 5 && !terminalThread.isRunning(); i++) {
```

Comment:
What happens if it is still not running after 5 loops ? We should consider to add timeout mechanism here and allow the timeout threshold configurable.

Reply:
Ok, I am adding a timeout error log.
But I want to change 5 to 10, wait a total of 5 seconds, it should be enough. Zeppelin has a lot of configuration items. right?